refactor: enable noImplicitOverride in project

devversion opened the pull request:
Enables the `noImplicitOverride` option to improve
code health in the repository.

Read more about this option here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#override-and-the---noimplicitoverride-flag

devversion committed:
build: enable noImplicitOverride in project

Enables the `noImplicitOverride` option to improve
overall codehealth in the repository.

devversion committed:
refactor(compiler): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `compiler` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(compiler-cli): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `compiler-cli` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(dev-infra): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `dev-infra` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(animations): ensure compatibility with noImplicitOverride

 Adds the `override` keyword to the `animations` sources to ensure
 compatibility with `noImplicitOverride`.

devversion committed:
refactor(core): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `core` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(platform-browser): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `platform-browser` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(platform-browser-dynamic): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `platform-browser-dynamic` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(platform-server): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `platform-server` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(router): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `router` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(upgrade): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `upgrade` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(benchpress): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `benchpress` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(common): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `common` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(elements): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `elements` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(forms): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `forms` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(service-worker): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `service-worker` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(language-service): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `language-service` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(localize): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `localize` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor(zone.js): ensure compatibility with noImplicitOverride

Adds the `override` keyword to the `zone.js` sources to ensure
compatibility with `noImplicitOverride`.

devversion committed:
refactor: ensure compatibility with noImplicitOverride for examples

Adds the `override` keyword to the `examples` sources to ensure
compatibility with `noImplicitOverride`.

petebacondarwin commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -63,7 +63,7 @@ export interface CompilerFacade {
 
   FactoryTarget: typeof FactoryTarget;
-  ResourceLoader: {new(): ResourceLoader};
+  ResourceLoader: Function&{prototype: ResourceLoader};
```
This is an interesting change.
Can you give some background behind the reason for this and what the error was without it?

devversion replied:
This is necessary because `ResourceLoader` is now an abstract class and therefore not `new-able`. I'll add a comment.

petebacondarwin commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -1016,7 +1016,7 @@ function convertStmtIntoExpression(stmt: o.Statement): o.Expression|null {
 export class BuiltinFunctionCall extends cdAst.FunctionCall {
   constructor(
-      span: cdAst.ParseSpan, sourceSpan: cdAst.AbsoluteSourceSpan, public args: cdAst.AST[],
+      span: cdAst.ParseSpan, sourceSpan: cdAst.AbsoluteSourceSpan, args: cdAst.AST[],
```
This seems like an additional refactoring that is not relevant to `noImplicitOverride`?

devversion replied:
It's necessary because `public args` here overrides the existing `public args` member set by `cdAst.FunctionCall`:

```ts
export class FunctionCall extends AST {
  constructor(
      span: ParseSpan, sourceSpan: AbsoluteSourceSpan, public target: AST|null,
      public args: any[]) {
```

petebacondarwin commented on packages/compiler/src/expression_parser/ast.ts:
```diff
@@ -24,16 +24,16 @@ export class ParseSpan {
-    return null;
-  }
+
+  abstract visit(visitor: AstVisitor, context?: any): any;
```
So this is a little bit worrying. By making this abstract, the `noImplicitOverride` seems to have no effect on the method in the child class, which sort of defeats the purpose of this compiler setting right? I.e. if the base class removed the abstract method, there would be no indication to update the child class.

petebacondarwin replied:
I think it might be worth adding the override to the concrete override of this abstract method anyway?

devversion replied:
Hmm. This is a good point. At the same time though if we do not make this `abstract`, then derived classes will not get any indication if the `visit` method is not supplied, right? Using `override` for every "actual implementation" of `AST` seems a bit like misuse (and requires a lot of changes). It seems like ideally this would be an interface?

devversion replied:
@petebacondarwin Relevant discussion in the TS repo: https://github.com/microsoft/TypeScript/issues/44457

What do you think about adding a custom TSLint rule that enforces we add `override` to abstract classes? I've tried setting one up, and it's not too much work, especially with a TSLint auto-fixup.

petebacondarwin replied:
FWIW I disagree with @RyanCavanaugh in https://github.com/microsoft/TypeScript/issues/44457#issuecomment-856170640.

It is generally unlikely that a new subclass would mistakenly implement the wrong method, since one would expect that the new code would be tested.

The real value of `override` is when the base class changes or removes the member that is being overridden.

petebacondarwin replied:
I'm not sure that adding to tslint at this stage in its life is a good use of our time.

devversion replied:
yeah, I agree. FWIW: The tslint rule is generic enough that it can be converted easily to ESLint or a standalone script, but at the current time it seems best to keep this a TSLint for the best integration with developer tools I think.

petebacondarwin commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -367,7 +367,7 @@ class TemplateBinder extends RecursiveAstVisitor implements Visitor {
    * at 1.
    */
-  static apply(template: Node[], scope: Scope): {
+  static applyWithScope(template: Node[], scope: Scope): {
```
Is this because the "class" is really a `Function` which has the built-in `apply()` method?

devversion replied:
yeah, exactly. I was considering adding a code comment, but as soon as this change lands, it's probably not needed because switching back to `apply` would cause a compile error then.

petebacondarwin commented on packages/compiler/src/resource_loader.ts:
```diff
@@ -7,11 +7,9 @@
- * An interface for retrieving documents by URL that the compiler uses
- * to load templates.
+ * An interface for retrieving documents by URL that the compiler uses to load
+ * templates. This is an abstract class so that it can be used as injection token.
```
Conventionally, it is better to put this sort of implementation node in a separate paragraph, saving the first paragraph as a description of what the class does/is.

petebacondarwin replied:
```suggestion
 * templates. This is an abstract class, rather than an interface, so that it can be used as injection token.
```

devversion replied:
👍

petebacondarwin commented on packages/compiler/testing/src/resource_loader_mock.ts:
```diff
@@ -17,7 +17,7 @@ export class MockResourceLoader extends ResourceLoader {
   private _requests: _PendingRequest[] = [];
 
-  get(url: string): Promise<string> {
+  override get(url: string): Promise<string> {
```
Interestingly, `ResourceLoader` is now abstract, along with `get()` so this override does not strictly appear to be necessary?

petebacondarwin replied:
Although I think it is probably good practice to always add `override` to all concrete methods that  implement abstract base methods.

devversion replied:
It's a mistake that this one still has `override`. I assume this happened because I first didn't consider making `ResourceLoader` abstract. Regarding, adding `override` to `abstract` members: I'm not sure about it; seems reasonable but at the same time I could imagine it being fine without `override` given there being another diagnostic if e.g. `get` is renamed accidentally.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/util/src/visitor.ts:
```diff
@@ -45,10 +45,8 @@ export abstract class Visitor {
-      VisitListEntryResult<ts.Statement, ts.ClassDeclaration> {
-    return {node};
-  }
+  abstract visitClassDeclaration(node: ts.ClassDeclaration):
+      VisitListEntryResult<ts.Statement, ts.ClassDeclaration>;
```
Nice!

petebacondarwin commented on packages/animations/browser/src/dsl/animation_timeline_builder.ts:
```diff
@@ -825,18 +825,18 @@ class SubTimelineBuilder extends TimelineBuilder {
 
   constructor(
-      driver: AnimationDriver, public element: any, public keyframes: ɵStyleData[],
+      driver: AnimationDriver, element: any, public keyframes: ɵStyleData[],
```
👍

petebacondarwin commented on packages/core/test/render3/jit/declare_factory_spec.ts:
```diff
@@ -52,7 +52,7 @@ class DependingClass {
 
 class ChildClass extends DependingClass {
-  static ɵfac =
+  static override ɵfac =
```
What does it mean to override a static property?
This doesn't seem to make sense to me. Static properties are not inherited, so this should have no effect, right?

petebacondarwin replied:
I see that this causes TS to error: [playground](https://www.typescriptlang.org/play?noImplicitOverride=true#code/MYGwhgzhAEBCkFNoG8BQ1oQC5iwS2GgDMB7EgCgEoUBfVO1USGAYQAs8QATaBADywIAdlxjwISNBmy4CxMlVr0gA)

But I think this is really a bug in the diagnostic, no?

devversion replied:
Unless I miss something, static properties are inherited, so this error makes sense?

```js
class A {
  static hello = "true";
}

class B extends A {}

B.hello;
```

petebacondarwin replied:
OMG! I never knew this.

So I see that with downleveled code (to ES5) TypeScript actually replaces the `B.hello` with `A.hello`.

petebacondarwin commented on packages/language-service/ivy/template_target.ts:
```diff
@@ -142,11 +142,22 @@ export interface TwoWayBindingContext {
+ * Special marker AST that can be used when the cursor is within the `sourceSpan` but not
+ * the key or value span of a node with key/value spans.
+ */
+class OutsideKeyValueMarkerAst extends e.AST {
+  visit(): null {
+    return null;
+  }
+}
```
Is this because you made `AST` abstract?

mary-poppins commented:
You can preview 3207044 at https://pr42512-3207044.ngbuilds.io/.
You can preview 9fff281 at https://pr42512-9fff281.ngbuilds.io/.
You can preview 68a462f at https://pr42512-68a462f.ngbuilds.io/.
You can preview 89cc905 at https://pr42512-89cc905.ngbuilds.io/.
You can preview 0ce1b47 at https://pr42512-0ce1b47.ngbuilds.io/.
You can preview b3d0d38 at https://pr42512-b3d0d38.ngbuilds.io/.

petebacondarwin reviewed (approved):
Great work @devversion!

Keeping our code clean: FTW!

devversion committed:
feat(dev-infra): add lint rule to enforce no-implicit-override abstract members

TypeScript introduced a new flag called `noImplicitOverride` as part
of TypeScript v4.3. This flag introduces a new keyword called `override`
that can be applied to members which override declarations from a base
class. This helps with code health as TS will report an error if e.g.
the base class changes the method name but the override would still
have the old method name. Similarly, if the base class removes the method
completely, TS would complain that the memeber with `override` no longer
overrides any method.

A similar concept applies to abstract methods, with the exception that
TypeScript's builtin `noImplicitOverride` option does not flag members
which are implemented as part of an abstract class. We want to enforce
this as a best-practice in the repository as adding `override` to such
implemented members will cause TS to complain if an abstract member is
removed, but still implemented by derived classes.

More details: https://github.com/microsoft/TypeScript/issues/44457.

devversion committed:
refactor: add override keyword to members implementing abstract declarations

In combination with the TS `noImplicitOverride` compatibility changes,
we also want to follow the best-practice of adding `override` to
members which are implemented as part of abstract classes. This
commit fixes all instances which will be flagged as part of the
custom `no-implicit-override-abstract` TSLint rule.

mary-poppins commented:
You can preview 5eef61a at https://pr42512-5eef61a.ngbuilds.io/.

mary-poppins commented:
You can preview b67029b at https://pr42512-b67029b.ngbuilds.io/.

josephperrott commented on dev-infra/tslint-rules/noImplicitOverrideAbstractRule.ts:
```diff
@@ -0,0 +1,141 @@
+import * as ts from 'typescript';
+
+const FAILURE_MESSAGE = 'Missing override modifier. Members implemented as part of ' +
+    'abstract classes should explicitly set the "override" modifier. ' +
```
```suggestion
    'abstract classes must explicitly set the "override" modifier. ' +
```

josephperrott commented on dev-infra/tslint-rules/noImplicitOverrideAbstractRule.ts:
```diff
@@ -0,0 +1,141 @@
+
+const FAILURE_MESSAGE = 'Missing override modifier. Members implemented as part of ' +
+    'abstract classes should explicitly set the "override" modifier. ' +
+    'More details: https://github.com/microsoft/TypeScript/issues/44457#issuecomment-856202843.';
```
I am a bit torn on instructing people to  read the comment for context.  Would it be better to express the reasoning here instead?

devversion replied:
That is something I was unsure about. Usually other lint rules also don't provide any context on the "why", and it is probably resulting in a long message/TSLint popup in IDE's, so I thought I'd link to the Github comment for details.

(although I'd have preferred linking to an actual error page like tsetse would do)

josephperrott replied:
Not a huge preference.  Lets leave it this way and worst case we find its annoying to have the complete link and we can remove it or change it

josephperrott commented on dev-infra/tslint-rules/noImplicitOverrideAbstractRule.ts:
```diff
@@ -0,0 +1,141 @@
+  }
+}
+
+function visitNode(node: ts.Node, ctx: WalkContext, program: ts.Program) {
+  // If a class element implements an abstract member but does not have the
+  // `override` keyword, create a lint failure.
```
```suggestion

/**
 * For a node and each of its child nodes, check whether the node is a class element which
 *  implements an abstract member but does not have the `override` keyword.
 */
function visitNode(node: ts.Node, ctx: WalkContext, program: ts.Program) {
```

josephperrott commented on dev-infra/tslint-rules/noImplicitOverrideAbstractRule.ts:
```diff
@@ -0,0 +1,141 @@
+
+/** Checks if the given class inherits an abstract member with the specified name. */
+function checkClassForInheritedMatchingAbstractMember(
+    clazz: ts.ClassDeclaration, typeChecker: ts.TypeChecker, searchMemberName: string): boolean {
```
Just wanted to note that I appreciate this name to avoid the keyword `class`

josephperrott reviewed (approved):
LGTM

Though we should be sure to have others review as you made a couple concrete classes abstract

IgorMinar commented on packages/service-worker/worker/testing/fetch.ts:
```diff
@@ -99,7 +99,6 @@ export class MockHeaders implements Headers {
 export class MockRequest extends MockBody implements Request {
   readonly isHistoryNavigation: boolean = false;
   readonly isReloadNavigation: boolean = false;
-  readonly body!: ReadableStream;
```
oops :)

devversion replied:
Yeah, the `MockBody` base class already defines the property equally, so we do not need to override anything.

IgorMinar commented on packages/service-worker/worker/testing/scope.ts:
```diff
@@ -38,7 +38,7 @@ export class WindowClientImpl extends MockClient implements WindowClient {
   frameType: ClientFrameType = 'top-level';
   url = 'http://localhost/unique';
 
-  constructor(readonly id: string) {
```
It's odd that Typescript allows this. It appears that while the compiler doesn't complain about it, it also doesn't enforce that the argument is readonly. Did this surface to you as a compiler error, or you just noticed it through visual inspection of the code?

devversion replied:
It caused an error with `noImplicitOverride` because `readonly id: string` overwrites the public `id` member from the base class. So to fix this, I just removed `readonly` here so that `id` becomes local here, and then when I pass the value to `super`, the base class exposes it as the public `id` property, if that makes sense. e.g.

```ts
class Base {
  constructor(readonly id: string) {}
}

class Derived {
  // id here does not need to be readonly/public as it would then override `id` from `Base`.
  constructor(id: string, otherParam: any) {
    super(id);
  }
}
```

IgorMinar replied:
makes sense! thanks for the clarification.

IgorMinar reviewed (approved):
LGTM! Thanks for taking this on @devversion !

AndrewKushnir reviewed (approved):
👍 

Reviewed-for: fw-forms

AndrewKushnir commented:
[Presubmit](http://test/OCL:384267294:BASE:384267498:1626111777158:9f48e8cf).

AndrewKushnir commented:
FYI, presubmits are successful for the changes in this PR.

josephperrott reviewed (approved):
LGTM


Reviewed-for: global-approvers
